### PR TITLE
http: changing filter interfaces to make downstream access explicit.

### PR DESCRIPTION
**Pull request by alyssawilk**
as a follow-up to https://github.com/envoyproxy/envoy/pull/22918, splitting downstream-only filter calls into an explicit API, so it's clear which functions are not safe for upstream filters to call.

When merging this PR, folks implementing C++ filters may have to update their filter code.   Any filter calling the following functions:
route(const Router::RouteCallback& cb)
setRoute(Router::RouteConstSharedPtr route)
clearRouteCache()
requestRouteConfigUpdate(RouteConfigUpdatedCallbackSharedPtr route_config_updated_cb)

will need to add downstreamCallbacks()-> in front of the relevant function call.

for unit tests, the same work will need to be done for direct function calls, or downstream_callbacks_ when interacting with mocks.

Risk Level: low
Testing: updated tests
Docs Changes: n/a
Release Notes: n/a?

### CODE REVIEW

File: source/common/router/router.cc [hunk 1]
```diff
@@ -1727,7 +1727,9 @@ bool Filter::convertRequestHeadersForInternalRedirect(Http::RequestHeaderMap& do
   }
   downstream_headers.setPath(path_and_query);
 
-  callbacks_->clearRouteCache();
+  if (callbacks_->downstreamCallbacks()) {
+    callbacks_->downstreamCallbacks()->clearRouteCache();
+  }
```
Comment (mattklein123) [inline source/common/router/router.cc:1732]: Isn't this guaranteed to be non-null? If not can you add a comment on why?

**Review by mattklein123 — CHANGES_REQUESTED** [review]
Thanks makes sense. Can you also add a release note? Nice!

/wait

**Comment by alyssawilk** [comment]
done.  I'd initially gone to add a release note and there really isn't an appropriate area - it's not a behavioral change but it's the closest category I guess?

**Review by mattklein123 — APPROVED** [review]
Thanks!

**Pull request merged**